src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -545,37 +547,115 @@ where
             Poll::Pending => source.block_until_ready()?,
         }
     };
-    match deps.iter().map(|p| p.package_id()).max() {
-        Some(pkgid) => {
-            let pkg = Box::new(source).download_now(pkgid, config)?;
-            Ok(pkg)
+    if let Some(pkgid) = deps.iter().map(|p| p.package_id()).max() {
+        let pkg = Box::new(source).download_now(pkgid, config)?;
+        Ok(pkg)
+    } else {
+        let spe = SelectPackageError::new(dep, source);
+        bail!(anyhow::Error::from(spe))
+    }
+}
+
+struct YankedInfo {
+    package_name: InternedString,
+    source_id: SourceId,
+}
+
+struct PackageNotFound {
+    package_name: InternedString,
+    source_id: SourceId,
+    // REVIEW make it a reference or is cloning ok?
+    version_req: OptVersionReq,
+}
+
+enum SelectPackageError {
```

Comment:
How come `SelectPackageError` was created rather than just using `bail`?

Reply:
Personally, I found the if/else branches using bail hard to read and repetitive. 

I performed the refactor to familiarise myself with the codebase. 

after I finished it, adding an enum variant converting it to an anyhow::Error was quite easy to make the test pass

Reply:
Since this error enum and its conversion to anyhow::Error is private to this module, I thought the benefits of defining new readable types would outweigh the PR review costs

Reply:
Unless I'm missing something, this looks like it overcomplicates the code.  I'm not seeing us getting any type-safety, programmatic error case checking, or code reuse benefits.  We've replaced simple one-line bail's with 3 types and a `From` impl.

> I thought the benefits of defining new readable types would outweigh the PR review costs

Please keep these things separate.  If nothing else, at least separate commits.

Reply:
> Unless I'm missing something, this looks like it overcomplicates the code. I'm not seeing us getting any type-safety, programmatic error case checking, or code reuse benefits. We've replaced simple one-line bail's with 3 types and a `From` impl.

I can separate the logic that fixes the test into 1 commit and my final refactor into another. Then you can review them separately and if you want only the test fix, I will revert the refactor commit. 

> Please keep these things separate. If nothing else, at least separate commits.

will do

Reply:
> Then you can review them separately and if you want only the test fix, I will revert the refactor commit.

While I can see value to splitting out the error analysis into a separate function, I feel the more specific error types are boiletplate without benefits to justify it.

Reply:
Thanks for the input. I will revert the refactoring commit and continue working on the smaller PR surface.